Support extended arguments to task pool deserialization.

scott-gibson-sociomantic opened the pull request:
The main driver behind this change is to support passing runtime-dependant information
(such as DHT, DMQ instances) into the deserialized tasks as previously the only way this
could be done was to create static member variables which isn't ideal.

I've also included a small refactoring of the ThrottledTaskPool to simplify the code as the overriden
start/restore are no longer needed and can be done in an overriden startImpl

release notes coming _soon_.

Geod24 commented on src/ocean/task/TaskPool.d:
```diff
@@ -180,33 +180,33 @@ class TaskPool ( TaskT : Task ) : ObjectPool!(Task)
         ***********************************************************************/
 
-        public bool restore ( void[] serialized )
+        public bool restore ( Args ... ) ( Args args )
```
That looks wrong. This method is `override`n in a derived class, but it's a templated method. I'm surprised the compiler does not complain.

E.g. the following code triggers an assertion failure:
```d
class Foo
{
    void restore (Args...) (Args a) { assert(0); }
}

class Bar : Foo
{
    override void restore (Args...) (Args a) {}
}

void main ()
{
    scope bar = new Bar();
    boom(bar);
}

void boom (scope Foo f) { f.restore(); }
```

scott-gibson-sociomantic replied:
My original reasoning for changing this to Args... instead of using something like ParametersOf!(TaskT.deserialize) was to support multiple types of function signatures. (IIRC using ParametersOf! means only the first declaration is used. Perhaps the proposed change causes more problems that i didn't anticipate. I removed the overwritten "restore" from the ThrottledTaskPool and i could set this as final but that's technically a api change then no?

Geod24 replied:
Yes it's a breaking change. You can actually generate an overload set matching another overload set, but it's a PITA. Want the code ?

matthias-wende-sociomantic commented:
What about adding a new templated restore method with a different name, that is final and can then be used to implement the original (virtual) restore method.

scott-gibson-sociomantic committed:
Refactor ThrottledTaskPool to remove overrides of start()/restore()

These methods were duplicating the existing methods but adding method calls
for triggering suspension. this was in the past because provided tasks
were overwritten as the task system didn't support termination hooks.

Now that the derived internal class is removed these methods can be
refactored away and the logic can be done in an override of startImpl.

scott-gibson-sociomantic committed:
Support extra args in task pool deserialization.

When restoring disk, there was no way to restore runtime information
that is unable to be written to disk (eg. objects, function pointers).
In the past the suggested method was to have a static set method
applicable to all tasks but this change allows the user to pass additional arguments to the
deserialize method.

scott-gibson-sociomantic commented:
I've modified this to instead of using templated arguments to use ParametersOf!(TaskT.deserialize) as well as adding extra templated methods of TaskPoolSerializer.load in case someone inherited it. I believe this makes it have no breaking changes.

scott-gibson-sociomantic commented:
Apologies some changes (TaskPool.d) was missing in my last push. Updated
